src/OrbitBase/WhenAll.cpp

```diff
@@ -12,15 +12,18 @@
 #include "absl/synchronization/mutex.h"
 
 namespace orbit_base {
-Future<void> JoinFutures(absl::Span<const Future<void>> futures) {
+Future<void> WhenAll(absl::Span<const Future<void>> futures) {
```

Comment:
Let me know what you think about the name. Is it too generic?

I'm also considering moving all the executor/promise/future related stuff into a separate module/namespace (`orbit_executors`). Would it still be too generic in a separate namespace?

Reply:
Before I can answer this question, I have another question: would `WhenAny` take multiple "regular" (not timeout) futures? In that case, how would you distinguish which one completed in case the type is the same? And also (my ignorance), what would happen to the ones that have not completed?

Reply:
The return type of WhenAny is a Future that contains a std::variant which can have duplicate types. `std::variant<int, int, int>` is totally fine. In that case you would determine which future returned by its index.

The ones that did not complete will be ignored/lost if you don't have a separate copy of that future. If you are referring to the task that is supposed to complete the future: It is not touched at all. The ThreadPool (or any other Executor) won't be notified that their task lost the race somewhere down the stream.

Reply:
> In that case you would determine which future returned by its index.

Aaaand I now see how it was a stupid question.

So back to the original questions.

> Let me know what you think about the name. Is it too generic?

 I do think that the name is generic. But I also don't really have problems with it, "when" makes me think of the "future" I guess.

To make it a bit less generic I can propose `JoinAllFutures` and `JoinAnyFuture`. But maybe you already considered something like that and there's a reason why "when" is semantically better than "join".

Another alternative could be to keep the "when" but add "futures": `WhenAllFutures`, `WhenAnyFuture`. Again, maybe "when ... future" is semantically incorrect?

> Would it still be too generic in a separate namespace?

It wouldn't. But I don't know if it's really necessary to move to a different module.

Reply:
How about keeping `JoinFutures` and adding `WhenAny`? 
On a low level, these functions might be quite similar, but from a high level perspective, they do something quite different. `JoinFutures` feels right to me, because its not changing the individual futures (no continuation is cancelled). It "just" transforms an array of futures to a future of array. `WhenAny` is quite different, in that the user will only ever get the result of one of the futures. 

So why align the names at all? To have more elegant code (whatever that means)? I think readability/expressiveness is the most important thing here. If I would not know about `WhenAny`, would I change the name from `JoinFutures` to `WhenAll`?.. probably not.

I also do not feel strongly about this at all

Reply:
As discussed offline. But for the sake of completeness here my thinking:

I'm pretty convinced that `All` and `Any` are the right terms here for a couple of reasons.

1. `Join` to me means that we make one future out of many futures. Both the proposed `WhenAny` and `WhenAll` do that but `Join` doesn't imply how they are merged - so keeping that for `WhenAll` could be confusing.
2. `All` and `Any` are used in other languages as well. Example JS: `Promise.all` and `Promise.any`
3. `All` and `Any` are also used for similar fold-like operations in C++. See `std::any_of` and `std::all_of` for example.

So my conclusion is that there is no major push back against the names `WhenAll` and `WhenAny`. I think I prefer to keep them because I think they are most idiomatic (Even though I quite like Davide's suggestion of `WhenAllFutures`.)